### PR TITLE
Enable linux runtime helper for linux function apps

### PR DESCRIPTION
**Pull request by Flanker32**
1. Update function.json format to fit both windows and linux
2. Enable linux runtime helper for linux function apps

Todo:
Add test cases for function runtime handlers

### CODE REVIEW

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AbstractFunctionMojo.java [hunk 1]
```diff
@@ -152,6 +153,18 @@ public FunctionApp getFunctionApp() throws AzureAuthFailureException {
         return null;
     }
 
+    public RuntimeConfiguration getRuntime() {
+        return runtime;
+    }
+
+    public OperatingSystemEnum getRuntimeOs() throws MojoExecutionException {
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AbstractFunctionMojo.java]: why throw MojoExecutionException  in a get method?

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/handlers/runtime/FunctionRuntimeHandler.java [hunk 2]
```diff
@@ -37,6 +36,15 @@ protected FunctionRuntimeHandler(Builder<?> builder) {
...
     }
 
+    @Override
+    public abstract FunctionApp.DefinitionStages.WithCreate defineAppWithRuntime() throws Exception;
+
+    @Override
+    public abstract FunctionApp.Update updateAppRuntime(FunctionApp app) throws Exception;
+
+    @Override
+    public abstract AppServicePlan updateAppServicePlan(FunctionApp app) throws Exception;
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/handlers/runtime/FunctionRuntimeHandler.java]: avoid throw general Exception: https://stackoverflow.com/questions/7959461/throwing-generic-exception-discouraged

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 3]
```diff
@@ -166,6 +152,30 @@ protected void configureAppSettings(final Consumer<Map> withAppSettings, final M
 
     //endregion
 
+    protected FunctionRuntimeHandler getFunctionRuntimeHandler() throws MojoExecutionException, AzureAuthFailureException {
+        final FunctionRuntimeHandler.Builder<?> builder;
+        switch (this.getRuntimeOs()) {
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java]: remove getRuntimeOs but add a function here to get os from runtime

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/PackageMojo.java [hunk 4]
```diff
@@ -250,7 +252,10 @@ protected void writeObjectToFile(final ObjectWriter objectWriter, final Object o
             throws IOException {
         targetFile.getParentFile().mkdirs();
         targetFile.createNewFile();
-        objectWriter.writeValue(targetFile, object);
+        final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        objectWriter.writeValue(byteArrayOutputStream, object);
+        // Change line of end to LF, refers https://stackoverflow.com/questions/3776923/how-can-i-normalize-the-eol-character-in-java
+        FileUtils.write(targetFile, byteArrayOutputStream.toString().replaceAll("\\r\\n?", "\n"));
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/PackageMojo.java]: use a more elegant way of generating json file with correct line feeds.

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AbstractFunctionMojo.java (hunk 1 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AbstractFunctionMojo.java]: We need to parse string to enum, which will throw MojoExecutionException

Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AbstractFunctionMojo.java]: Removed this method and change to get os from `RuntimeConfiguration`.

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/handlers/runtime/FunctionRuntimeHandler.java (hunk 2 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/handlers/runtime/FunctionRuntimeHandler.java]: Done

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java (hunk 3 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java]: Done

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/PackageMojo.java (hunk 4 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/PackageMojo.java]: Change to use `DefaultIndenter` to do this job

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 5]
```diff
@@ -83,42 +86,27 @@ protected void doExecute() throws Exception {
 
     protected void createOrUpdateFunctionApp() throws Exception {
         final FunctionApp app = getFunctionApp();
+        final FunctionRuntimeHandler runtimeHandler = getFunctionRuntimeHandler();
         if (app == null) {
-            createFunctionApp();
+            createFunctionApp(runtimeHandler);
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java]: use getFunctionRuntimeHandler() inside createFunctionApp

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/PackageMojo.java [hunk 6]
```diff
@@ -254,9 +257,11 @@ protected void writeObjectToFile(final ObjectWriter objectWriter, final Object o
     }
 
     protected ObjectWriter getObjectWriter() {
+        final DefaultIndenter indenter = DefaultIndenter.SYSTEM_LINEFEED_INSTANCE.withLinefeed("\n");
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/PackageMojo.java]: use interface, not impl class

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/configurations/RuntimeConfiguration.java [hunk 7]
```diff
@@ -20,6 +27,10 @@ public void setOs(String os) {
         this.os = os;
     }
 
+    public OperatingSystemEnum getOperationSystemEnum() throws MojoExecutionException {
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/configurations/RuntimeConfiguration.java]: move this mothod to the place where to use  it, change MojoExecutionException at OperatingSystemEnum.fromString to  IllegalArgumentException

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java (hunk 5 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java]: Done

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/PackageMojo.java (hunk 6 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/PackageMojo.java]: Changed to `DefaultPrettyPrinter.Indenter`

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/configurations/RuntimeConfiguration.java (hunk 7 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/configurations/RuntimeConfiguration.java]: Moved to `Utils` in maven-plugin-lib

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 8]
```diff
@@ -90,35 +93,21 @@ protected void createOrUpdateFunctionApp() throws Exception {
         }
     }
 
-    protected void createFunctionApp() throws Exception {
+    protected void createFunctionApp() throws MojoExecutionException, AzureAuthFailureException {
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java]: AzureAuthFailureException, MojoExecutionException

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java [hunk 9]
```diff
@@ -39,7 +41,7 @@ public String getOs() {
 
     public OperatingSystemEnum getOsEnum() {
         try {
-            return OperatingSystemEnum.fromString(this.os);
+            return Utils.parseOperationSystem(this.os);
```
Comment (Flanker32) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/RuntimeSetting.java]: Todo: Remove this method

**Pull request merged**